firedrake/linear_solver.py

```diff
@@ -41,153 +31,59 @@ def __init__(self, A, *, P=None, solver_parameters=None,
-        self.ksp.setDM(W.dm)
-        self.ksp.setDMActive(False)
-
-        if nullspace is not None:
-            nullspace._apply(self.A)
-            if P is not None:
-                nullspace._apply(self.P)
-
-        if transpose_nullspace is not None:
-            transpose_nullspace._apply(self.A, transpose=True)
-            if P is not None:
-                transpose_nullspace._apply(self.P, transpose=True)
-
-        if near_nullspace is not None:
-            near_nullspace._apply(self.A, near=True)
-            if P is not None:
-                near_nullspace._apply(self.P, near=True)
-
-        self.nullspace = nullspace
-        self.transpose_nullspace = transpose_nullspace
-        self.near_nullspace = near_nullspace
-        # Operator setting must come after null space has been
-        # applied
-        self.ksp.setOperators(A=self.A.petscmat, P=self.P.petscmat)
-        # Set from options now (we're not allowed to change parameters
-        # anyway).
-        self.set_from_options(self.ksp)
-
-    @cached_property
-    def test_space(self):
-        return self.A.arguments()[0].function_space()
-
-    @cached_property
-    def trial_space(self):
-        return self.A.arguments()[1].function_space()
-
-    @cached_property
-    def _ulift(self):
-        return function.Function(self.trial_space)
-
-    @cached_property
-    def _blift(self):
-        return cofunction.Cofunction(self.test_space.dual())
-
-    @cached_property
-    def _update_rhs(self):
-        from firedrake.assemble import get_assembler
-        expr = -action(self.A.a, self._ulift)
-        # needs_zeroing = False adds -A * ulift to the exisiting value of the output tensor
-        # zero_bc_nodes = True writes the BC data on the boundary nodes of the output tensor
-        return get_assembler(expr, bcs=self.A.bcs, zero_bc_nodes=False, needs_zeroing=False).assemble
-
-    def _lifted(self, b):
-        self._ulift.dat.zero()
-        for bc in self.A.bcs:
-            bc.apply(self._ulift)
-        # Copy the input to the internal Cofunction so we do not modify the input
-        self._blift.assign(b)
-        self._update_rhs(tensor=self._blift)
-        # self._blift is now b - A u_bc, and satisfies the boundary conditions
-        return self._blift
+        self.ksp = self.snes.ksp
 
     @PETSc.Log.EventDecorator()
     def solve(self, x, b):
         """Solve the linear system with RHS ``b`` and store the solution in ``x``.
 
         Parameters
         ----------
-        x : firedrake.function.Function or firedrake.vector.Vector
-            Existing Function or Vector to place the solution to the linear system in.
-        b : firedrake.cofunction.Cofunction or firedrake.vector.Vector
-            A Cofunction or Vector with the right-hand side of the linear system.
+        x : firedrake.function.Function
+            Existing Function to place the solution to the linear system in.
```

Comment:
Just to be consistent with the docstring for `b`. I don't think it needs to be stated that `x` needs to exist!
```suggestion
            A Function to place the solution to the linear system in.
```